rockcraft/oci.py

```diff
@@ -189,6 +200,83 @@ def add_layer(
 
+    def add_user(
+        self,
+        prime_dir: Path,
+        base_layer_dir: Path,
+        tag: str,
+        username: str,
+        uid: int,
+    ) -> None:
+        """Create a new ROCK user.
+
+        :param prime_dir: Path to the user-defined parts' primed content.
+        :param base_layer_dir: Path to the base layer's root filesystem.
+        :param tag: The ROCK's image tag.
+        :param username: Username to be created. Same as group name.
+        :param uid: UID of the username to be created. Same as GID.
+        """
+        # pylint: disable=too-many-arguments
+        user_files = {"passwd": "", "group": "", "shadow": ""}
+
+        prime_dir_etc = prime_dir / "etc"
+        base_layer_dir_etc = base_layer_dir / "etc"
+        # Being cautious about possible changes (edits or removals) in
+        # /etc/{passwd,group,shadow} done by the user through overlay scripts.
+        # Basically:
+        #  - if it exists in prime, use it,
+        #  - if it doesn't exist in prime AND isn't "whiteout", use the base,
+        #  - if it is "whiteout" or doesn't exist anywhere, use an empty file.
+        # NOTE: "shadow" is only modified it it already exists.
+        for u_file in user_files:
+            if (prime_dir_etc / u_file).exists():
+                user_files[u_file] = (prime_dir_etc / u_file).read_text()
+            elif (base_layer_dir_etc / u_file).exists() and not (
+                prime_dir_etc / f".wh.{u_file}"
+            ).exists():
+                user_files[u_file] = (base_layer_dir_etc / u_file).read_text()
+
+        if (
+            f"{username}:" in user_files["passwd"]
+            or f":{uid}:" in user_files["passwd"]
+            or f"{username}:" in user_files["group"]
+            or f":{uid}:" in user_files["group"]
+        ):
+            raise errors.RockcraftError(
+                str(
+                    f"Error while trying to create user {username}:{uid}, "
+                    f"with group {username}:{uid}...\n"
+                    " - conflict with existing user/group in the base filesystem"
+                )
+            )
+
+        user_files[
+            "passwd"
+        ] += f"{username}:x:{uid}:{uid}::/{Pebble.PEBBLE_PATH}:/usr/bin/false\n"
+        user_files["group"] += f"{username}:x:{uid}:\n"
+
+        with tempfile.TemporaryDirectory() as tmpfs:
+            tmpfs_etc = Path(tmpfs) / "etc"
+            tmpfs_etc.mkdir(parents=True, exist_ok=True)
+            with open(tmpfs_etc / "passwd", "a+") as passwdf:
+                passwdf.write(user_files["passwd"])
+
+            with open(tmpfs_etc / "group", "a+") as groupf:
+                groupf.write(user_files["group"])
+
+            if user_files["shadow"]:
+                days_since_epoch = (datetime.utcnow() - datetime(1970, 1, 1)).days
+
+                # only add the shadow file if there's already one in the base image
+                with open(tmpfs_etc / "shadow", "a+") as shadowf:
```

Comment:
Do we need specific permissions on this file?

Reply:
we only append something to this file if it already exists, and afaik the existing file permissions are preserved, so we shouldn't need to touch that.